addons/l10n_it_edi/tests/test_ir_mail_server.py

```diff
@@ -0,0 +1,142 @@
+        xmlns:p="http://ivaservizi.agenziaentrate.gov.it/docs/xsd/fatture/v1.2" 
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
+        xsi:schemaLocation="http://ivaservizi.agenziaentrate.gov.it/docs/xsd/fatture/v1.2 http://www.fatturapa.gov.it/export/fatturazione/sdi/fatturapa/v1.2/Schema_del_file_xml_FatturaPA_versione_1.2.xsd">
+          <FatturaElettronicaHeader>
+            <CessionarioCommittente>
+              <DatiAnagrafici>
+                <CodiceFiscale>01234560157</CodiceFiscale>
+              </DatiAnagrafici>
+            </CessionarioCommittente>
+          </FatturaElettronicaHeader>
+          <FatturaElettronicaBody>
+            <DatiGenerali>
+              <DatiGeneraliDocumento>
+                <TipoDocumento>TD02</TipoDocumento>
+              </DatiGeneraliDocumento>
+            </DatiGenerali>
+          </FatturaElettronicaBody>
+        </p:FatturaElettronica>"""
+
+    @classmethod
+    def setUpClass(cls):
+        """ Setup the test class with a PEC mail server and a fake fatturaPA content """
+
+        super().setUpClass(chart_template_ref='l10n_it.l10n_it_chart_template_generic',
+                           edi_format_ref='l10n_it_edi.edi_fatturaPA')
+
+        # Initialize the company's codice fiscale
+        cls.company = cls.company_data['company']
+        cls.company.l10n_it_codice_fiscale = 'IT01234560157'
+
+        # Build test data.
+        # invoice_filename1 is used for vendor bill receipts tests
+        # invoice_filename2 is used for vendor bill tests
+        cls.invoice_filename1 = 'IT01234567890_FPR01.xml'
+        cls.invoice_filename2 = 'IT01234567890_FPR02.xml'
+        cls.invoice_content = cls._get_test_content(cls.invoice_filename1)
+        cls.invoice = cls.env['account.move'].create({
+            'move_type': 'in_invoice',
+            'ref': '01234567890'
+        })
+        cls.attachment = cls.env['ir.attachment'].create({
+            'name': cls.invoice_filename1,
+            'datas': base64.encodebytes(cls.invoice_content),
+            'res_id': cls.invoice.id,
+            'res_model': 'account.move',
+        })
+        cls.edi_document = cls.env['account.edi.document'].create({
+            'edi_format_id': cls.edi_format.id,
+            'move_id': cls.invoice.id,
+            'attachment_id': cls.attachment.id,
+            'state': 'sent'
+        })
+
+        # Initialize the fetchmail server that has to be tested
+        cls.server = cls.env['fetchmail.server'].sudo().create({
+            'name': 'test_server',
+            'server_type': 'imap',
+            'l10n_it_is_pec': True})
+
+    @classmethod
+    def _get_test_content(cls, filename):
+        """ Get the content of a test file inside this module """
+        path = get_module_resource('l10n_it_edi', 'tests', 'expected_xmls', filename)
+        with tools.file_open(path, mode='rb') as test_file:
```

Comment:
Unfortunately, copying code is not a guarantee that you're following best practices, as the level of reviews and expected quality is increasing :-)

Directly using `open()` or `io.open()` is now forbidden, you must always go through `tools.file_open()`. And it should automatically support any relative paths inside the server's `addons_path`, so `get_module_resource()` is redundant:

```suggestion
        path = 'l10n_it_edi/tests/expected_xmls/' + filename
        with tools.file_open(path, mode='rb') as test_file:
```

(You could use `os.path.join()` to construct the path, but that may trigger `ci/security` too, and you don't need it)

Reply:
| Unfortunately, copying code is not a guarantee that you're following best practices, as the level of reviews and expected quality is increasing :-)
Not a guarantee, but certainly it's a place to start, unless it really stinks. 
I generally read code written by others before coming up with mine, as a rule of thumb.
So, nice to hear that the focus on code quality is increasing. 

Yet, I cannot say I find string concatenation an attractive solution for building paths, but I see the point and will fix.

Reply:
> Yet, I cannot say I find string concatenation an attractive solution for building paths, but I see the point and will fix.

As I told you, feel free to use `os.path.join`